### PR TITLE
cleanup `glusterd2.socket` during start

### PR DESCRIPTION
**Pull request by aravindavk**
Used the lock file to find any other glusterd2 running with same
socket file in use.

Fixes: #1246
Signed-off-by: Aravinda VK <avishwan@redhat.com>

### CODE REVIEW

File: glusterd2/servers/sunrpc/server.go [hunk 1]
```diff
@@ -53,6 +55,25 @@ var clientsList = struct {
...
 
 	f := path.Join(config.GetString("rundir"), gd2SocketFile)
+	gd2LockFile := f + ".lock"
+	fd, err := syscall.Open(gd2LockFile,
+		syscall.O_CREAT|syscall.O_WRONLY|syscall.O_CLOEXEC, 0666)
+	if err != nil {
+		log.WithError(err).WithField("lockfile", gd2LockFile).Fatal("failed to open lock file")
+	}
+
+	defer syscall.Close(fd)
```
Comment (prashanthpai) [inline glusterd2/servers/sunrpc/server.go]: Won't closing unlock the file ? Shouldn't we be holding the lock through-out the lifetime of the process ?

**Comment by atinmu** [comment]
LGTM

File: glusterd2/servers/sunrpc/server.go (hunk 1 above)
Comment (prashanthpai) [reply glusterd2/servers/sunrpc/server.go]: I'd prefer if the fd is closed in `Stop()` method after we close the socket file listener. You can store the fd in `SunRPC` structure.

Testcase for this doesn't make much sense and may be really hard even if possible.

However, I'd test this manually by bringing up multiple gd2 instances. One way I could think of:

1. Brick up one gd2 instance and keep it running.
2. Hardcode socket file path to the one used by first instance and recompile gd2
3. Bring up the 2nd instance. This should fail.

Comment (aravindavk) [reply glusterd2/servers/sunrpc/server.go]: Fixed

**Comment by aravindavk** [comment]
Tested manually as follows

Test 1 to see stale socket file will not cause failure on restart

```
[root@fedora28 ~]# systemctl start glusterd2

[root@fedora28 ~]# ps -ax | grep glusterd2
 3825 ?        Ssl    0:00 /usr/sbin/glusterd2 --config=/etc/glusterd2/glusterd2.toml
 3844 pts/0    S+     0:00 grep --color=auto glusterd2

[root@fedora28 ~]# ls -l /var/run/glusterd2/glusterd2.socket*
srwxr-xr-x. 1 root root 0 Oct  4 15:48 /var/run/glusterd2/glusterd2.socket
-rw-r--r--. 1 root root 0 Oct  4 14:02 /var/run/glusterd2/glusterd2.socket.lock

[root@fedora28 ~]# kill -9 3825

[root@fedora28 ~]# ls -l /var/run/glusterd2/glusterd2.socket*
srwxr-xr-x. 1 root root 0 Oct  4 15:48 /var/run/glusterd2/glusterd2.socket
-rw-r--r--. 1 root root 0 Oct  4 14:02 /var/run/glusterd2/glusterd2.socket.lock

[root@fedora28 ~]# systemctl restart glusterd2

[root@fedora28 ~]# ls -l /var/run/glusterd2/glusterd2.socket*
srwxr-xr-x. 1 root root 0 Oct  4 15:49 /var/run/glusterd2/glusterd2.socket
-rw-r--r--. 1 root root 0 Oct  4 14:02 /var/run/glusterd2/glusterd2.socket.lock
```

Test2: To see the lock is held till glusterd2 is running

```
[root@fedora28 ~]# ps -ax | grep glusterd2                                                                                                                   
 3863 ?        Ssl    0:00 /usr/sbin/glusterd2 --config=/etc/glusterd2/glusterd2.toml                                                                        
 3885 pts/0    S+     0:00 grep --color=auto glusterd2

[root@fedora28 ~]# python
>>> import fcntl
>>> f = open("/var/run/glusterd2/glusterd2.socket.lock")
>>> fcntl.flock(f.fileno(), fcntl.LOCK_EX|fcntl.LOCK_NB)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
IOError: [Errno 11] Resource temporarily unavailable

[root@fedora28 ~]# systemctl stop glusterd2
[root@fedora28 ~]# python
>>> import fcntl
>>> f = open("/var/run/glusterd2/glusterd2.socket.lock")
>>> fcntl.flock(f.fileno(), fcntl.LOCK_EX|fcntl.LOCK_NB)
>>>
```

File: glusterd2/servers/sunrpc/server.go [hunk 2]
```diff
@@ -53,6 +56,23 @@ var clientsList = struct {
...
+	if err != nil {
+		log.WithError(err).WithField("lockfile", gd2LockFile).Fatal("failed to open lock file")
+	}
+
+	err = syscall.Flock(fd, syscall.LOCK_EX|syscall.LOCK_NB)
+	if err != nil {
+		log.WithError(err).WithField("socket", gd2SocketFile).Fatal("failed to get lock")
+	}
+
+	err = os.Remove(f)
```
[the remark this answers is not in the document]
Comment (prashanthpai) [reply glusterd2/servers/sunrpc/server.go]: Right! Oops

**Pull request merged**